storage/integrate.go

```diff
@@ -0,0 +1,337 @@
+// to leak memory.
+type TreeBuilder struct {
+	getTile getPopulatedTileFunc
+	rf      *compact.RangeFactory
+}
+
+// NewTreeBuilder creates a new instance of TreeBuilder.
+//
+// The getTile param must know how to fetch the specified tile from storage. It must return an instance of os.ErrNotExist
+// (either directly, or wrapped) if the requested tile was not found.
+func NewTreeBuilder(getTile func(ctx context.Context, tileID TileID, treeSize uint64) (*api.HashTile, error)) *TreeBuilder {
+	readCache := newTileReadCache()
+	getFullTile := func(ctx context.Context, tileID TileID, treeSize uint64) (*populatedTile, error) {
+		r, ok := readCache.Get(tileID, treeSize)
+		if ok {
+			return r, nil
+		}
+		t, err := getTile(ctx, tileID, treeSize)
+		if err != nil {
+			return nil, err
+		}
+		ft, err := newPopulatedTile(t)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create fulltile: %v", err)
+		}
+		readCache.Set(tileID, treeSize, ft)
+		return ft, nil
+	}
+	return &TreeBuilder{
+		getTile: getFullTile,
+		rf:      &compact.RangeFactory{Hash: rfc6962.DefaultHasher.HashChildren},
+	}
+}
+
+// newRange creates a new compact.Range for the specified treeSize, fetching tiles as necessary.
+func (t *TreeBuilder) newRange(ctx context.Context, treeSize uint64) (*compact.Range, error) {
+	rangeNodes := compact.RangeNodes(0, treeSize, nil)
+	errG := errgroup.Group{}
+	hashes := make([][]byte, len(rangeNodes))
+	for i, id := range rangeNodes {
+		i := i
+		id := id
+		errG.Go(func() error {
+			tLevel, tIndex, nLevel, nIndex := layout.NodeCoordsToTileAddress(uint64(id.Level), id.Index)
+			ft, err := t.getTile(ctx, TileID{Level: tLevel, Index: tIndex}, treeSize)
+			if err != nil {
+				return err
+			}
+			h := ft.Get(compact.NodeID{Level: nLevel, Index: nIndex})
+			if h == nil {
+				return fmt.Errorf("missing node: [%d/%d@%d]", id.Level, id.Index, treeSize)
+			}
+			hashes[i] = h
+			return nil
+		})
+	}
+	if err := errG.Wait(); err != nil {
+		return nil, err
+	}
+	return t.rf.NewRange(0, treeSize, hashes)
+}
+
+func (t *TreeBuilder) Integrate(ctx context.Context, fromSize uint64, entries [][]byte) (newSize uint64, rootHash []byte, tiles map[TileID]*api.HashTile, err error) {
+	baseRange, err := t.newRange(ctx, fromSize)
+	if err != nil {
+		return 0, nil, nil, fmt.Errorf("failed to create range covering existing log: %w", err)
+	}
+
+	// Initialise a compact range representation, and verify the stored state.
+	r, err := baseRange.GetRootHash(nil)
+	if err != nil {
+		return 0, nil, nil, fmt.Errorf("invalid log state, unable to recalculate root: %w", err)
+	}
+
+	klog.V(1).Infof("Loaded state with roothash %x", r)
+	// Create a new compact range which represents the update to the tree
+	newRange := t.rf.NewEmptyRange(fromSize)
+	if len(entries) == 0 {
+		klog.V(1).Infof("Nothing to do.")
+		// Nothing to do, nothing done.
+		return fromSize, r, nil, nil
+	}
```

Comment:
I don't know how common this will be, but I'd stick it at the top of the method to save all the work you don't need to do.

Reply:
It should be pretty uncommon. Unfortunately, it'll be a little more work to move it all the way up since we return the root hash; we'd have to return some sort of sentinel error or similar to signal.

Not sure if it's worth it at the mo'?

Reply:
I'd missed that it returned `r`. No worries.